src/ucp/api/ucp.h

```diff
@@ -333,22 +335,102 @@ typedef enum {
- * within a single data transfer function call.
+ * within a single data transfer function call. Each buffer is described by
+ * the data type of a single element and the amount of consecutive elements.
+ * A data type can be a simple contiguous buffer, or a complex layout such as
+ * a strided data type - which also allows for interleaving of items from
+ * multiple such data types.
+ *
+ * For example, a simple IOV can be two disjoint buffers, of arbitrary length:
+ *
+ * iov[0].buffer = ptr1;
+ * iov[0].dt = ucp_dt_make_contig(4);
+ * iov[0].count = 1;
+ * iov[1].buffer = ptr2;
+ * iov[1].dt = ucp_dt_make_contig(1);
+ * iov[1].count = 4;
+ *
+ * - In this case, both buffers are of size 4.
  *
- * @note If @a length is zero, the memory pointed to by @a buffer
+ * @note If @a count is zero, the memory pointed to by @a buffer
  *       will not be accessed. Otherwise, @a buffer must point to valid memory.
  */
 typedef struct ucp_dt_iov {
-    void   *buffer;   /**< Pointer to a data buffer */
-    size_t  length;   /**< Length of the @a buffer in bytes */
+    void           *buffer; /**< Pointer to a data buffer */
+    ucp_datatype_t dt;      /**< Type of the data buffer */
+    size_t         count;   /**< Amount of items in the data buffer */
 } ucp_dt_iov_t;
 
 
 /**
  * @ingroup UCP_DATATYPE
+ * @brief Structure for a single dimension of a strided data type.
+ *
+ * This structure is used to specify an item in a list of dimensions, used to
+ * describe the requested layout of a strided data type.
+ */
+typedef struct ucp_dt_strided_dim {
+    size_t extent;
+    size_t count;
+} ucp_dt_strided_dim_t;
+
+
+/**
+ * @ingroup UCP_DATATYPE
+ * @brief Generate a reusable identifier for a strided data type.
+ *
+ * This macro creates an identifier for a data type consisting of multiple
+ * identical items with a fixed interval between each two adjacent ones.
+ * The resulting identifier needs to be released using @ref ucp_dt_destroy .
+ *
+ * @param [in]  datatype          Datatype descriptor for a single element.
+ * @param [in]  dimensions        Layout description per dimension of striding.
+ * @param [in]  dimension_count   Amount of dimensions describing the layout.
+ * @param [in]  interleave_ratio  For interleaving - amount of elements used in
+ *                                each cycle from this datatype (0 otherwise).
+ * @param [in]  is_reusable       Whether the datatype stores extra meta-data.
+ *
+ * A "Reusable" identifier means that the user can apply this identifier to
+ * multiple calls requiring datatypes, to benefit from optimizations on
+ * recurring access. This is done by storing additional meta-data, such as
+ * memory mapping info, to be used in subsequent calls.
+ * Limitations apply: after the first use of a reusable identifier - it can
+ * only be used with pointers to the same memory regions - in the same order.
+ *
+ * @return Data-type identifier.
+ */
+ucp_datatype_t ucp_dt_make_strided(ucp_datatype_t datatype,
```

Comment:
generic is actually a specific type, where user provides the callbacks - we called it "generic", but in most cases we don't want to trouble the user to implement callbacks. If there was one general-purpose dt-creating function - it would probably have some "switch-case" inside to implement each dt-type... in your question - do you mean this kind of function? 

I just followed notation to add this new strided type, but we can certainly use the opportunity to unify creation functions (currently macros, actually).